File: src/semantic_project_model/graph.py
"""Small directed graph algorithms."""

from __future__ import annotations

from collections.abc import Iterable


def adjacency(nodes: Iterable[str], edges: Iterable[tuple[str, str]]) -> dict[str, set[str]]:
    graph: dict[str, set[str]] = {node: set() for node in nodes}
    for source, target in edges:
        graph.setdefault(source, set()).add(target)
        graph.setdefault(target, set())
    return graph


def find_cycle(graph: dict[str, set[str]]) -> tuple[str, ...] | None:
    visiting: set[str] = set()
    visited: set[str] = set()
    stack: list[str] = []

    def visit(node: str) -> tuple[str, ...] | None:
        if node in visited:
            return None
        if node in visiting:
            start = stack.index(node)
            return (*stack[start:], node)
        visiting.add(node)
        stack.append(node)
        for target in sorted(graph.get(node, set())):
            cycle = visit(target)
            if cycle is not None:
                return cycle
        stack.pop()
        visiting.remove(node)
        visited.add(node)
        return None

    for node in sorted(graph):
        cycle = visit(node)
        if cycle is not None:
            return cycle
    return None


def topological_order(graph: dict[str, set[str]]) -> tuple[str, ...]:
    indegree = dict.fromkeys(graph, 0)
    for targets in graph.values():
        for target in targets:
            indegree[target] += 1

    ready = sorted(node for node, degree in indegree.items() if degree == 0)
    order: list[str] = []
    while ready:
        node = ready.pop(0)
        order.append(node)
        for target in sorted(graph[node]):
            indegree[target] -= 1
            if indegree[target] == 0:
                ready.append(target)
                ready.sort()

    if len(order) != len(graph):
        raise ValueError("graph is cyclic")
    return tuple(order)


def longest_path(graph: dict[str, set[str]], weights: dict[str, int]) -> tuple[str, ...]:
    order = topological_order(graph)
    score = {node: weights.get(node, 1) for node in graph}
    previous: dict[str, str | None] = dict.fromkeys(graph)

    for source in order:
        for target in graph[source]:
            candidate = score[source] + weights.get(target, 1)
            if candidate > score[target]:
                score[target] = candidate
                previous[target] = source

    if not score:
        return ()
    end = max(score, key=lambda node: (score[node], node))
    path: list[str] = []
    cursor: str | None = end
    while cursor is not None:
        path.append(cursor)
        cursor = previous[cursor]
    return tuple(reversed(path))

File: src/project-model/graph.ts
export type Adjacency = ReadonlyMap<string, ReadonlySet<string>>;

export const adjacency = (
  nodes: Iterable<string>,
  edges: Iterable<readonly [string, string]>,
): Adjacency => {
  const graph = new Map<string, Set<string>>();
  for (const node of nodes) graph.set(node, new Set());
  for (const [source, target] of edges) {
    const targets = graph.get(source) ?? new Set<string>();
    targets.add(target);
    graph.set(source, targets);
    if (!graph.has(target)) graph.set(target, new Set());
  }
  return graph;
};

export const findCycle = (graph: Adjacency): ReadonlyArray<string> | undefined => {
  const visiting = new Set<string>();
  const visited = new Set<string>();
  const stack: Array<string> = [];

  const visit = (node: string): ReadonlyArray<string> | undefined => {
    if (visited.has(node)) return undefined;
    if (visiting.has(node)) {
      const start = stack.indexOf(node);
      return [...stack.slice(start), node];
    }
    visiting.add(node);
    stack.push(node);
    const targets = [...(graph.get(node) ?? [])].sort();
    for (const target of targets) {
      const cycle = visit(target);
      if (cycle !== undefined) return cycle;
    }
    stack.pop();
    visiting.delete(node);
    visited.add(node);
    return undefined;
  };

  for (const node of [...graph.keys()].sort()) {
    const cycle = visit(node);
    if (cycle !== undefined) return cycle;
  }
  return undefined;
};

export const topologicalOrder = (graph: Adjacency): ReadonlyArray<string> | undefined => {
  const indegree = new Map([...graph.keys()].map((node) => [node, 0]));
  for (const targets of graph.values()) {
    for (const target of targets) indegree.set(target, (indegree.get(target) ?? 0) + 1);
  }

  const ready = [...indegree]
    .filter(([, degree]) => degree === 0)
    .map(([node]) => node)
    .sort();
  const order: Array<string> = [];
  while (ready.length > 0) {
    const node = ready.shift()!;
    order.push(node);
    for (const target of [...(graph.get(node) ?? [])].sort()) {
      const degree = (indegree.get(target) ?? 0) - 1;
      indegree.set(target, degree);
      if (degree === 0) {
        ready.push(target);
        ready.sort();
      }
    }
  }
  return order.length === graph.size ? order : undefined;
};

export const longestPath = (
  graph: Adjacency,
  weights: ReadonlyMap<string, number>,
): ReadonlyArray<string> => {
  const order = topologicalOrder(graph);
  if (order === undefined) return [];
  const score = new Map([...graph.keys()].map((node) => [node, weights.get(node) ?? 1]));
  const previous = new Map<string, string | undefined>(
    [...graph.keys()].map((node) => [node, undefined]),
  );

  for (const source of order) {
    for (const target of graph.get(source) ?? []) {
      const candidate = (score.get(source) ?? 0) + (weights.get(target) ?? 1);
      if (candidate > (score.get(target) ?? 0)) {
        score.set(target, candidate);
        previous.set(target, source);
      }
    }
  }
  if (score.size === 0) return [];
  const end = [...score].sort(([leftId, left], [rightId, right]) =>
    left === right ? rightId.localeCompare(leftId) : right - left,
  )[0]![0];
  const path: Array<string> = [];
  let cursor: string | undefined = end;
  while (cursor !== undefined) {
    path.push(cursor);
    cursor = previous.get(cursor);
  }
  return path.reverse();
};
